### PR TITLE
QRDA Import date fixes and class fixes

### PR DESCRIPTION
**Pull request by sjpadgett**
fix careplan dates to return null when empty
fix negate for medication order
fix negated helper logic

### CODE REVIEW

**Comment by sjpadgett** [comment]
@kchapple note.

Once I figure out why cypress has me download a patient test deck of patients for import for a given measure and yet errors when I give the same patients back for test i.e gives 78 patients then says only 75 expected, we'll get green! 
However for the CMS22v10 we pass validation and calculations, yeah. May not seem like much but for us, a big deal!

image

**Comment by bradymiller** [comment]
2DV-2

**Comment by stephenwaite** [comment]
@sjpadgett , just let me know when you'd like me to test something please. I created a login on your cypress.

**Comment by sjpadgett** [comment]
I'll send you the test steps in email then follow the c(1) where you have to enter attributes into cypress form.
The test run like a check list. I haven't looked to closely at the test so it's an adventure we're both on!:)

**Comment by sjpadgett** [comment]
I'll set you up a product to use. Give me a few minutes

**Comment by sjpadgett** [comment]
@stephenwaite Set you up product Stephen W CQM

**Comment by sjpadgett** [comment]
Okay thanks but I'm really confused where the device template went. I rechecked cat3 and cat1 so should be good.

**Comment by sjpadgett** [comment]
sorry @kchapple missed concern on last commit. tired!

image

**Comment by sjpadgett** [comment]
@kchapple Decided to go a different way with encounter ends date. It gets blurry using calendar events relationship to encounter.
I added end date to form_encounter and modified the qdm encounter service to reflect.

**Comment by sjpadgett** [comment]
@kchapple Need service for careplan device_order

image

**Comment by sjpadgett** [comment]
have to add reason to procedure orders!!

**Comment by sjpadgett** [comment]
@kchapple I had to add reason code to procedure intervention preformed type that the ProcedureService uses.
Can you add reason to the service. I would add negation also on reason status though I haven't found any to this template in any patient samples, it is in the spec.

**Comment by sjpadgett** [comment]
Never mind on reason I was able to do it. Just check my commit for me to ensure I don't step on anything

**Comment by sjpadgett** [comment]
@kchapple This makes CMS22v10, CMS69v10 and CMS125v10 is a go. Next for me will be CMS122v10 but hooray for us.
Will merge this PR and start again with next testing hiccup!

**Comment by sjpadgett** [comment]
I've learned that we are expected to validate imports! nuts, how! Anyway using normal test where invalid reports slip by me I can get calculator errors where the report shouldn't have been submitted. So until I figure out I have to validate imports I am testing against cypress known good patients. e.g. Here is CMS122v10 normal test(this patient didn't have any encounters):

image

And with cypress known good patients:

image

So this is good news and bad news! Anyway, add CMS122v10 to list.

**Comment by sjpadgett** [comment]
And CMS127v10

image

**Comment by sjpadgett** [comment]
CMS130v10 Anyone getting bored seeing green?

image

**Comment by bradymiller** [comment]
go green!!!

https://www.youtube.com/watch?v=BVDnDORSY9E

File: sql/6_1_0-to-7_0_0_upgrade.sql [hunk 1]
```diff
@@ -315,3 +315,16 @@ ALTER TABLE `form_care_plan` CHANGE `date` `date` DATETIME NULL DEFAULT NULL;
...
 #EndIf
+
+#IfMissingColumn form_encounter date_end
+ALTER TABLE `form_encounter` ADD `date_end` DATETIME DEFAULT NULL;
+#EndIf
+
+#IfMissingColumn procedure_order_code date_end
+ALTER TABLE `procedure_order_code` ADD `date_end` datetime DEFAULT NULL;
+ALTER TABLE `procedure_order_code` ADD `reason_code` varchar(31) DEFAULT NULL;
+ALTER TABLE `procedure_order_code` ADD `reason_description` text DEFAULT NULL;
```
Comment (bradymiller) [inline sql/6_1_0-to-7_0_0_upgrade.sql:326]: remove the default null from text (here and in database.sql)

**Comment by sjpadgett** [comment]
Needs fixed in service. 

image

**Comment by sjpadgett** [comment]
@kchapple be sure to run sql upgrade procedure code size was to small for an oid

**Comment by sjpadgett** [comment]
@kchapple you'll need to add negation to immunization service while I go fix conflicts, again! I added reason code and status to table.

File: sql/6_1_0-to-7_0_0_upgrade.sql [hunk 2]
```diff
@@ -316,6 +316,203 @@ ALTER TABLE `form_care_plan` CHANGE `date` `date` DATETIME NULL DEFAULT NULL;
 ALTER TABLE api_token ADD COLUMN `context` TEXT COMMENT 'context values that change/govern how access token are used';
 #EndIf
 
+#IfRow2D list_options list_id language notes eng
+CREATE TEMPORARY TABLE lang_updates_610 (notes char(15), twodigit_notes char(2));
+START TRANSACTION;
+INSERT INTO lang_updates_610 VALUES
+('abk','ab'),
+('aar','aa'),
```
Comment (sjpadgett) [inline sql/6_1_0-to-7_0_0_upgrade.sql:324]: @bradymiller @adunsulag I had to fix conflicts so you guys need to ensure sql upgrade and database are still intact.
Looks okay to me but check me anyways. Thanks

File: sql/6_1_0-to-7_0_0_upgrade.sql [hunk 3]
```diff
@@ -511,4 +511,32 @@ DROP TABLE lang_updates_610;
...
+
+#IfNotColumnType procedure_order_code procedure_code VARCHAR(64)
+ALTER TABLE `procedure_order_code` CHANGE `procedure_code` `procedure_code` VARCHAR(64) NOT NULL DEFAULT '' COMMENT 'like procedure_type.procedure_code';
+#EndIf
+
+#IfNotColumnType procedure_order date_ordered DATETIME
+ALTER TABLE `procedure_order` CHANGE `date_ordered` `date_ordered` DATETIME DEFAULT NULL;
+#EndIf
+
+#IfMissingColumn immunizations reason_code
```
Comment (bradymiller) [inline sql/6_1_0-to-7_0_0_upgrade.sql:537]: immunizations changes are missing in database.sql

Comment (sjpadgett) [reply sql/6_1_0-to-7_0_0_upgrade.sql:537]: thanks

File: sql/database.sql [hunk 4]
```diff
@@ -9236,14 +9237,20 @@ DROP TABLE IF EXISTS `procedure_order_code`;
...
   `do_not_send`             tinyint(1)  NOT NULL DEFAULT '0'    COMMENT '0 = normal, 1 = do not transmit to lab',
   `procedure_order_title`   varchar( 255 ) NULL DEFAULT NULL,
   `procedure_type`          varchar(31) DEFAULT NULL,
   `transport`               varchar(31) DEFAULT NULL,
+  `date_end` datetime DEFAULT NULL,
+  `reason_code` varchar(32) DEFAULT NULL,
+  `reason_description` text,
+  `reason_date_low` datetime DEFAULT NULL,
+  `reason_date_high` datetime DEFAULT NULL,
+  `reason_status` varchar(32) DEFAULT NULL,
```
Comment (bradymiller) [inline sql/database.sql:9253]: varchar is 31 in upgrade

**Comment by sjpadgett** [comment]
gesh! what a marathon!
@kchapple immunization template is not seeing the oid valueset and sets negationInd to 'false'
The nulled cvx code is in diff component than we're used to seeing.

Will bring in so we can reset devs to see why you can't get diagnosis to populate issue_encounter. Maybe change to windows! lol

**Pull request merged**